tunnel/intra/protect/protect.go

```diff
@@ -0,0 +1,127 @@
+// Copyright 2019 The Outline Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package protect
+
+import (
+	"context"
+	"net"
+	"sync/atomic"
+	"syscall"
+)
+
+// Protector is a wrapper for Android's VpnService.protect().
+type Protector interface {
+	// Protect a socket, i.e. exclude it from the VPN.
+	// This is needed in order to avoid routing loops for the VPN's own sockets.
+	Protect(socket int32) bool
+}
+
+func makeRawConnControl(p Protector) (func(uintptr)) {
+	return func(fd uintptr) {
+		if p != nil {
+			if !p.Protect(int32(fd)) {
+				panic("Failed to protect socket")
+			}
+		}
+	}
+}
+
+func makeDialerControl(p Protector) (func(string, string, syscall.RawConn) error) {
+	rawConnControl := makeRawConnControl(p)
+	return func(network, address string, c syscall.RawConn) error {
+		return c.Control(rawConnControl)
+	}
+}
+
+func dialer(p Protector) *net.Dialer {
+	return &net.Dialer{
+		Control: makeDialerControl(p),
```

Comment:
Could this have a performance penalty when `p == nil`?

Reply:
I think the overhead should be minimal, something like three function calls and one if-statement.  This only happens once per socket, so it should be small compared to the amount of work we do for every packet.